docs/proposals/14-wireguard-based-cluster-vpn.md

```diff
@@ -0,0 +1,178 @@
+
+The WireGuard network needs a separate network range/cidr. It has to be unique for the seed and all its shoot clusters. An example for an assumed workload of around 250 shoot clusters with enough buffer would be 192.168.0.0/22 (1024 IP addresses), i.e. 192.168.0.0-192.168.3.255. The IP addresses from this range need to be managed, but the IP address management (IPAM) using the gardener Kubernetes objects like seed and shootstate as backing store is fairly straight forward. This is especially true as we do not expect large network ranges and only infrequent IP allocations. Hence, the IP address allocation can be quite simple, i.e. scan the range for a free IP address.
+
+There is another restriction: in case shoot clusters are configured to be seed clusters this network range must not overlap with the "parent" seed cluster. If the parent seed cluster uses `192.168.0.0/22` the child seed cluster can for example use `192.168.4.0/22`. Grandchildren can however use grandparent IP address ranges.
+
+This slightly adds to the restrictions described in the current solution outline. In that the arbitrary chosen `192.168.123.0/24` range is resticted. For the purpose of this implementation we propose to extend that restriction to `192.168.128.0/17` range. Most of it would be reserved for "future use however".
+
+WireGuard can utilize the Linux kernel so that after initialization/configuration no user space processes are required. We propose to recommend the wireguard kernel module as the default solution for all seeds. For shoot clusters, the WireGuard kernel based approach is also recommended, but the user space solution should also work as we expect less traffic on the shoot side. We expect the userspace implementation to work on all operating systems supported by Gardener in case no kernel module is available.
+
+There is a user space implementation of WireGuard, which can be used on Linux distributions without the WireGuard kernel module. (WireGuard moved into the standard Linux kernel 5.6.) Our proposal can handle the kernel/user space switch transparently, i.e. we include the user space binaries and use them only when required. However, especially for the seed the kernel based solution might be more attractive. Garden Linux 318.4.0 supports WireGuard.
+
+We have looked at Ubuntu and SuSE chost:
+
+- SuSE chost does not provide the WireGuard kernel module and it is not installable via zypper. It should however be straightforward for SuSE to include that in their next release.
+- Ubuntu does not provide the kernel module either but it can be installed using `apt-get install wireguard`. With that it appears straightforward to provide an image with WireGuard pre-installed.
```

Comment:
See above, this additional "management" effort is undesirable, and especially, there might operating systems we don't know of (everything is open-source and extensible), so it's not enough to look at the three known ones. 

Also, can you explain who/how these modules would be installed in your proposal above?

Reply:
Difficult. The WireGuard module is available for all Kernels >= 2.6. Since 5.6 it is part of the official Kernel.

I do believe we can assume the kernel module is available for all modern Linux distributions but the way it is installed will quite certainly differ. What we have seen with SuSE and Ubuntu is that it is not installed by default but can be installed using zypper or apt-get. I assume this is the same for RHEL, just that the install command is yum.

Reply:
OK, but who installs them and when? Gardener or the OS extension? Or must it be "pre-installed" in the image? How is backwards-compatibility ensured (let's say, we run a new Gardener version with this new WireGuard solution but an old image or old OS extension which does not (have) install(ed) the kernel module)? Or is this not a problem because it's "optional" anyways and we don't need to know whether it's present or not?

Generally, we should prevent installing packages on the fly as the nodes might run in environments without internet access. That's also the reason why we use a Docker image to "install" the kubelet/kubectl binaries on the worker nodes (we can assume that there is access to a container registry).

Reply:
> Generally, we should prevent installing packages on the fly as the nodes might run in environments without internet access.

For fenced environments, I guess the environment owner is providing the OS image and will be responsible to either configure package repository or to pre-install the required packages in the image. With SuSE and Garden Linux the packages are pre-installed, but the upstream Ubuntu images are/were not pre-installed with docker, thus the os-extension installs it on the fly if missing.

But I agree, adding wireguard module as requirement to existing landscapes is something we do not want. 
It would be nice if the solution can fall back automatically to user space binary that can be installed in the same way as kubelet/kubectl.

PS. There is also Flatcar.

Reply:
We assume that the wireguard kernel module is already pre-installed. Hence, neither gardener nor the os extension would do the kernel module installation. If it is not pre-installed the tunnel solution should fall-back to the wireguard user space solution.
I agree that it makes the operation a lot easier if it works transparently. Therefore, I used the term recommendation and not requirement.

Reply:
OK, fair enough. It would still be interesting to compare some numbers, i.e., what is the concrete consequence (in performance?) if I run a seed cluster w/ 200 shoots with or without the kernel module. This would not only be helpful to better judge this proposal but also for the documentation for Gardener operators later on.

Reply:
As we have decided to go with a solution almost identical to the existing one WireGuard performance numbers are not longer relevant for this proposal. We will however tacke this for Kubelink (where we also research real HA approaches).